services/versioncache/versioncache.go

```diff
@@ -0,0 +1,281 @@
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/services/agents"
+)
+
+const (
+	serviceCheckInterval = 24 * time.Hour
+	minCheckInterval     = 5 * time.Second
+)
+
+//go:generate mockery -name=Versioner -case=snake -inpkg -testonly
+
+// Versioner contains method for retrieving versions of different software.
+type Versioner interface {
+	GetVersions(pmmAgentID string, softwares []agents.Software) ([]agents.Version, error)
+}
+
+// Service is responsible for caching service software versions in the DB.
+type Service struct {
+	db       *reform.DB
+	l        *logrus.Entry
+	v        Versioner
+	updateCh chan struct{}
+}
+
+// New creates new service.
+func New(db *reform.DB, v Versioner) *Service {
+	return &Service{
+		db:       db,
+		l:        logrus.WithField("component", "version-cache"),
+		v:        v,
+		updateCh: make(chan struct{}, 1),
+	}
+}
+
+func (s *Service) syncServices() error {
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		serviceType := models.MySQLServiceType
+		services, err := models.FindServices(tx.Querier, models.ServiceFilters{ServiceType: &serviceType})
+		if err != nil {
+			return err
+		}
+
+		serviceIDs := make(map[string]struct{}, len(services))
+		for _, s := range services {
+			serviceIDs[s.ServiceID] = struct{}{}
+		}
+
+		serviceVersions, err := models.FindServicesSoftwareVersions(tx.Querier, nil)
+		if err != nil {
+			return err
+		}
+
+		// remove services software versions from the cache which are no longer exist
+		for _, sv := range serviceVersions {
+			if _, ok := serviceIDs[sv.ServiceID]; !ok {
+				if err := models.DeleteServiceSoftwareVersions(tx.Querier, sv.ServiceID); err != nil {
+					return err
+				}
+			}
+		}
+
+		// add new services software versions to the cache
+		cacheServiceIDs := make(map[string]struct{}, len(serviceVersions))
+		for _, sv := range serviceVersions {
+			cacheServiceIDs[sv.ServiceID] = struct{}{}
+		}
+		for _, s := range services {
+			if _, ok := cacheServiceIDs[s.ServiceID]; !ok {
+				if _, err := models.CreateServiceSoftwareVersions(tx.Querier, models.CreateServiceSoftwareVersionsParams{
+					ServiceID:        s.ServiceID,
+					SoftwareVersions: []models.SoftwareVersion{},
+					NextCheckAt:      time.Now().UTC(),
+				}); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+type prepareResults struct {
+	ServiceID   string
+	CheckAfter  time.Duration
+	NeedsUpdate bool
+	AgentID     string
+}
+
+// prepareUpdateVersions checks if there is any service that needs software versions update in the cache and
+// shifts the next check time for this service.
+func (s *Service) prepareUpdateVersions() (*prepareResults, error) {
+	results := &prepareResults{CheckAfter: minCheckInterval}
+
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		limit := 1
+		servicesVersions, err := models.FindServicesSoftwareVersions(tx.Querier, &limit)
```

Comment:
you can use pointer.ToInt.. as in other places in our code base

Reply:
done